### DISCUSSION TITLE
Blank screen with tft 3.5 "ILI9486 / ILI9488 / HX8357D

### DISCUSSION THREAD

**Discussion by Tizfaver** [ImpulseAdventure/GUIslice#428] [General]
Hi I bought this display: [this](https://www.amazon.it/dp/B07PLQS8MT/ref=pe_24968671_487022601_TE_SCE_dp_1)
In the back of the shiled is written that thew drivers are: ILI9488, ILI9486, HX8357D

I installed all libraries: Adafruit GFX, mcufriend, GUIslice-master

After this, I uncommented in `/src/GUIslice_config.h` the line that have `#include "../configs/ard-shld-mcufriend_4wire.h"`, and after that i changed in the file `/config/ard-shld-mcufriend_4wire.h` the line that have  `#define GSLC_SD_EN`    0 in   `#define GSLC_SD_EN    1`

Anrduino IDE run the code without problems, no error, nothing... the problem is that the screen is blank, and i don't see nothing. 
But with the example `GLUE_Demo_480x320` from the library MCUFRIEND_kbv the screen works.

Can somebody help me?

**Comment by Tizfaver** [comment]
Hi again, i tried again, but nothing... I don't know why my display only works with the `mcufriend_kbv` library, I can get almost all the sample skecths in the library to work. But it doesn't work with GUIslice generated code, why doesn't it use the `mcufriend` library?

in the serial monitor:

GUIslice version [0.17.0]:
ERROR: DrvInit() SD init failed
Init display handler [ADA_MCUFRIEND] FAIL
ERROR: Init() failed

**Comment by Pconti31** [comment]
Inside your ard-shld-mcufriend_4wire.h uncomment 
//. DBG_DRIVER
Then check your serial monitor for errors.

I suspect the board may not be id'ing itself correctly. If so at the top of your file is a line you can uncomment to your desired board I'd. Paul--

**Comment by ImpulseAdventure** [comment]
@Tizfaver -- the initialization is showing a failure because your SD card init failed.

Could you first let us know if the display initialization works if you set `#define GSLC_SD_EN 0`?
- If we still have problems, then it is possible that we need to force a MCUFRIEND ID. This can be done with the `DRV_DISP_ADAGFX_MCUFRIEND_FORCE` line near the top of your config file. The mcufriend example sketches sometimes do this for you automatically, so we may need to set it the same way here too to match.
- Thanks to Paul for the debug reminder -- we can enable more debug messaging if we set `#define DEBUG_ERR 2` as well.

**Reply by yesnoj** [reply]
Sorry for write here but...i'm using an Arduino 2, uncommented the ard-shld-mcufriend_xpt2046.h, updated the pins for 0x9340 :

  // MCUFRIEND_ID == 0x9340:
  // - DRV_TOUCH_ADA_SIMPLE [240x320]: (MCUFRIEND ID=0x9340) (XP=6,XM=A2,YP=A1,YM=7)  [TESTED]
  #define ADATOUCH_PIN_YP   A1
  #define ADATOUCH_PIN_XM   A2
  #define ADATOUCH_PIN_YM   7 
  #define ADATOUCH_PIN_XP   6 
  #define ADATOUCH_X_MIN    132
  #define ADATOUCH_X_MAX    885
  #define ADATOUCH_Y_MIN    837
  #define ADATOUCH_Y_MAX    150

I can see the buttons, but appears that the touch is not recognized...

I've leave the following :

  #define DRV_DISP_ADAGFX           // Adafruit-GFX library
  #define DRV_DISP_ADAGFX_MCUFRIEND // prenticedavid/MCUFRIEND_kbv
  //#define DRV_DISP_ADAGFX_MCUFRIEND_FORCE  0x9481 // MCUFRIEND write-only displays may need to force an ID value
  #define DRV_TOUCH_XPT2046_PS      // PaulStoffregen/XPT2046_Touchscreen

And changed the XPT2046_CS from 3 to A3...but i'm not sure i've done correctly.

  // Touch bus & pinout
  #define XPT2046_CS     A3
  #define XPT2046_IRQ    255 // IRQ pin or 255 if no IRQ used

I've tested the calibration, but nothing again.
With the MCUFriends calibration native sketch works perfectly : 

15:59:19.505 -> TouchScreen.h GFX Calibration
15:59:19.539 -> Not possible to diagnose Touch pins on ARM or ESP32
15:59:19.571 -> ID = 0x9340
15:59:42.391 -> 
15:59:42.391 -> cx=880 cy=242 cz=699  X, Y, Pressure
15:59:43.973 -> cx=521 cy=246 cz=602  X, Y, Pressure
15:59:45.107 -> cx=157 cy=239 cz=509  X, Y, Pressure
15:59:46.664 -> cx=865 cy=547 cz=636  X, Y, Pressure
15:59:48.168 -> cx=165 cy=553 cz=420  X, Y, Pressure
15:59:49.310 -> cx=868 cy=855 cz=634  X, Y, Pressure
15:59:50.170 -> cx=529 cy=850 cz=497  X, Y, Pressure
15:59:51.070 -> cx=156 cy=855 cz=301  X, Y, Pressure
15:59:51.428 -> 
15:59:51.428 -> *** COPY-PASTE from Serial Terminal:
15:59:51.534 -> const int XP=6,XM=A2,YP=A1,YM=7; //240x320 ID=0x9340
15:59:51.599 -> const int TS_LEFT=880,TS_RT=214,TS_TOP=135,TS_BOT=894;
15:59:51.632 -> 
15:59:51.632 -> PORTRAIT  CALIBRATION     240 x 320
15:59:51.698 -> x = map(p.x, LEFT=880, RT=214, 0, 240)
15:59:51.731 -> y = map(p.y, TOP=135, BOT=894, 0, 320)
15:59:51.764 -> 
15:59:51.764 -> LANDSCAPE CALIBRATION     320 x 240
15:59:51.796 -> x = map(p.y, LEFT=135, RT=894, 0, 320)
15:59:51.862 -> y = map(p.x, TOP=214, BOT=880, 0, 240)

Can you help me with this stuff?

**Comment by Pconti31** [comment]
@yesnoj Have you tried PaulStoffregen/XPT2046_Touchscreen/examples/TouchTest.ino or TouchTestIRQ.  Does either work?
if yes, post the constructor used and its pin definitions.
example:
```
#define CS_PIN  8
// MOSI=11, MISO=12, SCK=13

//XPT2046_Touchscreen ts(CS_PIN);
#define TIRQ_PIN  2
//XPT2046_Touchscreen ts(CS_PIN);  // Param 2 - NULL - No interrupts
//XPT2046_Touchscreen ts(CS_PIN, 255);  // Param 2 - 255 - No interrupts
XPT2046_Touchscreen ts(CS_PIN, TIRQ_PIN);  // Param 2 - Touch IRQ Pin - interrupt enabled polling```
```
Paul--

**Reply by yesnoj** [reply]
@Pconti31 thanks for your answer!
I don't know how set the pins are...i'm using the MCUFriend_kvb library and i've set these pins here :

//Touchscreen Calibrates value
#define SENSIBILITY 270
#define MINPRESSURE 500
#define MAXPRESSURE 1000

//These are the pins for the shield!Do calibration to set the correct pins!Otherwise touch WILL NOT WORK!Run Touchscreen_Native_Calib
#define YP A1
#define XM A2
#define YM 7 //YM><965 = TOUCHED!
#define XP 6

//Touchscreen calibration, Run Touchscreen_Native_Calib and set correct values
const uint16_t TS_LEFT = 132;//899;
const uint16_t TS_RT   = 885;//211;
const uint16_t TS_TOP  = 150;//940;
const uint16_t TS_BOT  = 837;//200;

// Init TouchScreen:
static TouchScreen_kbv ts(XP, YP, XM, YM, SENSIBILITY);
static TSPoint_kbv tp;

// Init LCD
static MCUFRIEND_kbv tft;

//LCD Orientation
static uint8_t swapXY = 1; //1: LANSCAPE 0:PORTRAIT

However if i run the script as you requested, this is a continuous output on serial :
17:45:28.106 -> Pressure = 4095, x = 8191, y = 8191

Thanks in advance for your help!

**Comment by Pconti31** [comment]
@yesnoj Its possible we have a language barrier here so I'll try again. 

Does  PaulStoffregen/XPT2046_Touchscreen/examples/TouchTest.ino  work? yes, no?

Previous post of yours says **With the MCUFriends calibration native sketch works perfectly** 
what program was that?
Arduino\libraries\MCUFRIEND_kbv\examples\diagnose_Touchpins.ino? yes, no?
TouchScreen_Calibr_native.ino? yes? no?

If you used diagnose_Touchpins.ino or TouchScreen_Calibr_native.ino successfully then your touch doesn't have a xpt2046 chip.  Its what is called a simple 4 pin touch which needs to use Arduino\libraries\Adafruit_TouchScreen library. 

If all of this is true first download this library and try its example
Arduino\libraries Adafruit_TouchScreen\examples\touchscreendemo\touchscreendemo.ino

edit it to use the pin outs that were reported by MCUFRIEND_kbv diagnose_Touchpins.ino
if  that works you can move on to guislice with a new config file configs/ard-shld-mcufriend_4wire.h

Paul--

**Reply by yesnoj** [reply]
Ciao Paul,
let me anwer to your question:

Does PaulStoffregen/XPT2046_Touchscreen/examples/TouchTest.ino work? _**No**_.
Arduino\libraries\MCUFRIEND_kbv\examples\diagnose_Touchpins.ino?  _**No**_.
TouchScreen_Calibr_native.ino? _**Yes**_.

Ok, so i don't have that chip :/ (xpt2046)...
Let me test it with the configs/ard-shld-mcufriend_4wire.h!
Thanks Paul!
I will update this post :)

**Reply by yesnoj** [reply]
Ciao Paul,
i've finally have it working!I've used the configs/ard-shld-mcufriend_4wire.h,
i've just changed the pins in the following way : 

  #define ADATOUCH_PIN_YP   A1
  #define ADATOUCH_PIN_XM   A2
  #define ADATOUCH_PIN_YM   7 
  #define ADATOUCH_PIN_XP   6 

And executed the script diag_ard_touch_calib.ino to have the exact values for the following :

  #define ADATOUCH_X_MIN    815
  #define ADATOUCH_X_MAX    162
  #define ADATOUCH_Y_MIN    151
  #define ADATOUCH_Y_MAX    878

I need now to understand if i can use these components into my already existing project.
I've wrote over 5000lines and i just need the slider components... :D
However, is possible with this GUIslice, implement a navigation between pages?
I've found this very difficult, but i've done...is a giant switch case in my project...but works
Thanks again Paul!

**Reply by yesnoj** [reply]
I'm doing some testing and i've found how go back and forward between the pages.
Really simple!Congratulation for this amazing tool!
But, i've found something not working , with the Spinner element.
I've added the spinner in my page and all works, until i interact with the spinner.
Once i click on it, nothing happens on spinner and i can't interact with other element anymore...i need to restart.
I've also tested the ex32_bld_spinner, with the same behavior....
In this example there is a progress bar that goes automatically from 0 to...200 and so on...but if i press the spinner...the bar stops his progress...
I've changed the value GSLC_FEATURE_COMPOUND to 1, as requested...but something is not workin...what i'm missing? :/

**Comment by Pconti31** [comment]
@yesnoj I suspect you are not using the latest version of guislice api since I repaired the spinner control over a year ago.  Issue #459 
open src/GUIslice_version.h it should show version number 
```
#define GUISLICE_VER "0.17.1.3"
```
If not you just need to update to latest version.  You can't use Arduino IDE for this you need to down load from github guislice repository on its main page you will see Code  click it and download the zip.

If you have 0.17.1.3 then I really have no idea what is going on.  

By the way, my 0.17.1.3 spinner doesn't need GSLC_FEATURE_COMPOUND so it can be set back to 0
Also, backup your configs/ard-shld-mcufriend_4wire.h, and GUIslice_config.h files before doing the unzip so you can just drop them back into the correct folders without needing to do any edits.
Paul--

**Reply by yesnoj** [reply]
Ciao Paul,
as you suspected, i've the #define GUISLICE_VER "0.17.0" . I've done the installation using the Arduino IDE, i will use the other approach via github as suggested!
Thanks!

PS. Do you think is possible to integrate just a component into my already existing project? 
Regards!

PPS. Just uptade to the latest release from github and works as expected! :)

**Comment by yesnoj** [comment]
@Pconti31 
I'm trying to implement some screens from my application, but just adding an image, bmp 24bit, cause the error that force to enable #define GSLC_SD_EN    1 .
If i set the value to 1, it compiles but unfortunally at sturtup remains completly white the screen.
Just removing the image and set back to 0, works as expected.
Can be solve this issue? 
I don't have so many images, i've used the method tft.drawRGBBitmap to print an image and works without any issue, but i can't do with GUISlice ...
Regards.

**Comment by Pconti31** [comment]
@yesnoj  **integrate just a component into my already existing project?**  
I would need a few more details to answer that question.

Did you create your project with the Builder?
If yes what version did you use?

If not, you created by hand then you are stuck with editing your project and adding new components by hand.  I had to remove importing non-builder projects due to unreasonable complains from users about so many fixes needing to be made after the import. 
It was designed to simple re-create the UI, components and their screen placement, and text and ignore any logic.   Meaning logic would need to be added to the new code skeleton.

Paul--

**Reply by yesnoj** [reply]
Paul i've created all my project by hand using the library :

#include <MCUFRIEND_kbv.h>
#include <TouchScreen_kbv.h>
#include <Adafruit_GFX.h>

All pages are created using the tft.fillScreen(BLACK) at the beginning, to empty the page...then adding new text and buttons passing coordinates and so on.
I will try to just add a checkbox, taken from GUISlicer, and trying to understand if i can interact with that.
Unfortunally i can't find cool stuff already available as a library that can be added to my little project...i really don't want to re-write all the stuff from the beginning...

**Comment by Pconti31** [comment]
@yesnoj **adding an image, bmp 24bit, cause the error**
The call tft.drawRGBBitmap doesn't use SD card it just uses a C array that stores an image.
You can do the same with GUIslice but you will need to make a small modification to your C array.
See example ex28_ard_btn_img_flash.ino and exit_n24.c
basically you edit your c array by adding height and width to the array instead of passing them through the api call.
```
// Please refer to https://github.com/ImpulseAdventure/GUIslice/wiki/Display-Images-from-FLAS
// for details on how to generate this file.

#include "GUIslice.h"
#include "GUIslice_config.h"
#if (GSLC_USE_PROGMEM)
  #if defined(__AVR__)
    #include <avr/pgmspace.h>
  #else
    #include <pgmspace.h>
  #endif
#endif

const unsigned short exit_n24[1024+2] GSLC_PMEM ={
32, // Height of image
32, // Width of image
```
than on edit your ino file to add your c array
```
extern "C" const unsigned short exit_n24[] PROGMEM;
```
and add a pointer to this array in your guislice ui element creationusing gslc_GetImageFromProg instead of gslc_GetImageFromSD
```
  pElemRef = gslc_ElemCreateBtnImg(&m_gui,E_ELEM_BTN_QUIT,E_PG_MAIN,(gslc_tsRect){258,70,32,32},
          gslc_GetImageFromProg((const unsigned char*)exit_n24,GSLC_IMGREF_FMT_BMP24),
```
Paul--

**Reply by yesnoj** [reply]
Thanks Paul, i will try it!
Anyway i've found a bug on the generation of the code.
If i crete some buttons and i want to make a copy of these in another pages, somethime the code isn't generated.
After the copy I've changed the ENUM to have always different buttons identification...but the problem remains.
If i create the buttons from scratch, than works always well.
I've deleted all the file except the .prj , and re-generated, but nothing changed...i need to create the buttons one-by-one ...is strange 
Regards.

**Reply by yesnoj** [reply]
> @yesnoj **adding an image, bmp 24bit, cause the error** The call tft.drawRGBBitmap doesn't use SD card it just uses a C array that stores an image. You can do the same with GUIslice but you will need to make a small modification to your C array. See example ex28_ard_btn_img_flash.ino and exit_n24.c basically you edit your c array by adding height and width to the array instead of passing them through the api call.
> 
> ```
> // Please refer to https://github.com/ImpulseAdventure/GUIslice/wiki/Display-Images-from-FLAS
> // for details on how to generate this file.
> 
> #include "GUIslice.h"
> #include "GUIslice_config.h"
> #if (GSLC_USE_PROGMEM)
>   #if defined(__AVR__)
>     #include <avr/pgmspace.h>
>   #else
>     #include <pgmspace.h>
>   #endif
> #endif
> 
> const unsigned short exit_n24[1024+2] GSLC_PMEM ={
> 32, // Height of image
> 32, // Width of image
> ```
> 
> than on edit your ino file to add your c array
> 
> ```
> extern "C" const unsigned short exit_n24[] PROGMEM;
> ```
> 
> and add a pointer to this array in your guislice ui element creationusing gslc_GetImageFromProg instead of gslc_GetImageFromSD
> 
> ```
>   pElemRef = gslc_ElemCreateBtnImg(&m_gui,E_ELEM_BTN_QUIT,E_PG_MAIN,(gslc_tsRect){258,70,32,32},
>           gslc_GetImageFromProg((const unsigned char*)exit_n24,GSLC_IMGREF_FMT_BMP24),
> ```
> 
> Paul--

Paul i've this error when i try to select the .c file generated  by your application GUIslice_Image2C
![image](https://github.com/ImpulseAdventure/GUIslice/assets/6757126/45704fca-0afb-46d5-91e8-439db72ec10a)
What i'm doing wrong?!

PS: Starting from an image available in the project ex28_ard_btn_img_flash.ino, i've copy-paste my array, and now works!
Maybe there was some issue during the export with your software...i don't know :/

**Comment by Pconti31** [comment]
@yesnoj post a zip file of your .prj file and I'll try to see the problem.

**Reply by yesnoj** [reply]
I will try to replicate the issue tomorrow, i'm done for today...
I've re-created all the flow in GUISlice, but of course is missing all the logic (and the logo image...i will try tomorrow your solution)!
Thanks Paul!

**Comment by Pconti31** [comment]
@yesnoj If you are using the builder and you have modified your C Image arrays to include height and width you can use the builder to create your image ui elements, just select the image ui element and the builder will first open arduino_res folder but you simply move to where ever you have stored your C file arrays (one file per image) and you will be able to select them and the builder will display and generate code for showing them.
Paul--

**Reply by yesnoj** [reply]
I will do tomorrow for sure!
A question, Is possible to have the "long" press on the buttons?

**Comment by Pconti31** [comment]
@yesnoj Depends upon what you mean by long press.  I have an example of a continuous press to run a motor while button is pressed then stop when not pressed inside my GUIsliceSolutions repository [apps/ContinuePress-Btn](https://github.com/Pconti31/GUIslice-Solutions/tree/main/apps/ContinuePress-Btn)  if this hepls you please leave a star in the repository.  I would like to have more than two stars here :)

If you need something else post a better description of your requirements.
Paul--

**Reply by yesnoj** [reply]
@Pconti31 the idea is this one... :)

https://github.com/ImpulseAdventure/GUIslice/assets/6757126/7af44c81-09f5-4911-ab31-d6d9a3c95b83

I've implemented on my project, maybe can be an idea for a future implementation :D

**Reply by yesnoj** [reply]
@Pconti31 
Some notes about GUIslice : 
1)Could be nice to re-order the pages, as is possible for the element inside the page itself. (also to have an incremental enum for each page), or at least to have the info about the ENUM value of the page in the info about the page...actually i can see just the ENUM var.
2)A bigger keypad/numpad, where is possible to set the size, the position, or if need to be full screen (like the one i've implemented in the short video above)
3)Have added for testing a keypad/numpad, but once i've deleted from the builder, the generated code still continues to add it...!
4) I'm trying to test what i need to convert my project, but i don't know why just change the color works, and not the text!
#define AVIO (gslc_tsColor){37, 150, 190}

gslc_tsElemRef* pElemRefTEST = gslc_PageFindElemById(&m_gui, E_PG_HOME, E_TXT_INFO2_SELECT_PROG);

gslc_ElemSetTxtStr(&m_gui,pElemRefTEST, (char*)"TEST");
gslc_ElemSetTxtCol(&m_gui,pElemRefTEST, AVIO);

I've tested the example ex04_ard_ctrls.ino where a button changes color and text once is pressed, and works of course...
In this example the button is created on code, not in a project...if i've understood correctly the logic: 

`  // Create Quit button with modifiable text label
  static char mstr_quit[8] = "Quit";
  pElemRef = gslc_ElemCreateBtnTxt(&m_gui,E_ELEM_BTN_QUIT,E_PG_MAIN,
    (gslc_tsRect){160,80,80,40},mstr_quit,sizeof(mstr_quit),E_FONT_BTN,&CbBtnQuit);
  gslc_ElemSetRoundEn(&m_gui,pElemRef,true);

and then...

  gslc_tsElemRef* pElemRef = gslc_PageFindElemById(&m_gui, E_PG_MAIN, E_ELEM_BTN_QUIT);

    // Change the button text
    gslc_ElemSetTxtStr(&m_gui, pElemRef, (char*)"STOP");`

**Comment by Pconti31** [comment]
_Some notes about GUIslice :
1)Could be nice to re-order the pages, as is possible for the element inside the page itself. (also to have an incremental enum for each page), or at least to have the info about the ENUM value of the page in the info about the page...actually i can see just the ENUM var._

It would be nice and I have thought about it but see my answer to your number 3. Remember though that the order you create pages has nothing to do with the order of display. You control that at runtime.

Also, my latest Builder shows you the page order in the tree view collapsed so even if you had 20 or more pages its easy to jump between them.

_2)A bigger keypad/numpad, where is possible to set the size, the position, or if need to be full screen (like the one i've implemented in the short video above)_
The API supports this and so does the Builder.
Inside builder on menubar edit->options select your keypad type by its tab: either KeyPadAlpha or KeyPadNum

For position change X and Y values.
Size of keypad change Use Default Button Sizes? to false then enter your new size for Button Width + Height.
You can also change fonts.

_3)Have added for testing a keypad/numpad, but once i've deleted from the builder, the generated code still continues to add it...!_

Yes, but without anyone in the community offering to help me out this won't get fixed.  You will need to delete by hand.
See my answer to GUIsliceBuilder [issue 47](https://github.com/ImpulseAdventure/GUIslice-Builder/issues/47)

_4) I'm trying to test what i need to convert my project, but i don't know why just change the color works, and not the text!_

You should read my FAQ.pdf.  See "Can you change a Text String at runtime?"

The reason I can't help more is that you failed to show the code creating your button.
Also, GUIslice is a static implementation.  I've made suggestions in the past to change this but for now we need to live with that. 

This means you can't add UI elements at runtime.  You need to allocate memory in advance. I don't know if that's your problem or not since again you haven't given enough code for me to see.
If the FAQ doesn't answer your question post a small demo project zip file and I'll check it out for you.

Paul--

**Reply by yesnoj** [reply]
@Pconti31 i've posted few lines of code of what i've tested to change the button text at runtime.
I'ts ok for me to have all the elements already on the screen, but if you tell me that i can't change the value of a text for a button or a simple text view, this can be a problem for me.

I'm trying now to make my keyboard also on GUISlice, bigger is better on a tiny screen as mine...but i'm not sure i will have the result i want...
The "TODO" string, need to change accordingly with my imput...
![image](https://github.com/ImpulseAdventure/GUIslice/assets/6757126/c1618279-8843-4ae1-a624-3899b750f6de)

PS. i can't find the FAQ.pdf!

**Comment by Pconti31** [comment]
@yesnoj For your long press see my GUIsliceSolutions repository [apps/Longpress-Btn](https://github.com/Pconti31/GUIsliceSolutions/tree/main/apps/Longpress-Btn)

The FAQ.pdf is with GUIsliceBuilder [Releases](https://github.com/ImpulseAdventure/GUIslice-Builder/releases/tag/v0.17.b22) as one of the 9 assets including the UserGuide.pdf

Again you only gave a screen shot no code.  I'm not a mind reader...

Post a zip of an actual tiny project file that fails.

Paul--

**Reply by yesnoj** [reply]
@Pconti31 i've read the file FAQ.pdf and i've solved just increasing the External Storage Size!
I've to understand where i can place the code for each case of the switch case automatically generated, because every time i auto-generate, the builder will delete my code :/

**Comment by Pconti31** [comment]
@yesnoj You turn off the generation of button callback edit->options General Tab at bottom
Preserve Button Callbacks? set to false;
Paul--

**Reply by yesnoj** [reply]
> @yesnoj You turn off the generation of button callback edit->options General Tab at bottom Preserve Button Callbacks? set to false; Paul--

I've set this to "true", i've read this also on the user_guide.pdf, but if i set true or false is the same...the code will be always wiped :/

![image](https://github.com/ImpulseAdventure/GUIslice/assets/6757126/086aa917-8678-407f-896c-ce093a5e2216)

I'm implementing the keyboard right now, trying to manage the buttons and strings : 

 case E_BTN_BACK_DEVELOPING:
        gslc_SetPageCur(&m_gui, E_PG_RESUME);
        break;
 case _1_BTN:
        gslc_ElemSetTxtStr(&m_gui, e_txt_keyboard_value, (char*)"1");
        break;

if i auto-generate this will be changed to the following :

 case E_BTN_BACK_DEVELOPING:
        gslc_SetPageCur(&m_gui, E_PG_RESUME);
        break;
 case _1_BTN:
        gslc_ElemSetTxtStr(&m_gui, e_txt_keyboard_value, (char*)"1");
        break;

will be preserved just what i've set using the builder

**Comment by Pconti31** [comment]
@yesnoj you can also avoid builder code by simply adding your own on the button property
Custom Code (optional) select the three ... 
and paste in your code without the break; or case statement which might be the best option for you.

also Preserve Button Callbacks? set to false; only works if you exit and reopen the builder.
Paul--

**Reply by yesnoj** [reply]
> @yesnoj you can also avoid builder code by simply adding your own on the button property Custom Code (optional) select the three ... and paste in your code without the break; or case statement which might be the best option for you.

> 
> also Preserve Button Callbacks? set to false; only works if you exit and reopen the builder. Paul--


Yep but i prefer to see the whole code how it also look :D
I've done a test, false or true, after the reboot of the builder , once i autogenerate , the code will be wiped :(

I've executed as "administrator" the software but nothing changes. Also if i set the preferred text color to BLACK, still remains yellow...even if in the preference page is changed to black...very strange :/

**Comment by Pconti31** [comment]
@yesnoj Without an actual example i can't help.
post zip of small project one or two UI elements along with exact steps one by one that causes the problem after setting  Preserve Button Callbacks? set to false;

Paul--

**Comment by Pconti31** [comment]
@yesnoj 
I forgot to remove customizing colors from edit->options->Text which I suspect you tried to use.

For color changes you can edit the GUIsliceBuilder/templates/guislice_themes.json file.
You can edit the first theme GUIslice, under 
```
			{
				"uiElemName": "Text",
				"frame_en": false,
				"fill_en": true,
				"text_col": "0xffff00",  // yellow (255,255,0)
				"frame_col": "0x808080", // GSLC_COL_GRAY (128,128,128) 
				"fill_col": "0x000000"   // black (0,0,0)
			},
```
and change any colors colors you want.  various web sites can convert rgb colors to hex values

You can also if add your own color theme if you understand json syntax otherwise just edit the GUIslice one.
After you edit, open Project options, select another theme, then go back and select GUIslice and it will rip through all previouly created UI elements and make them match your new colors.
Paul--

Paul--

**Comment by Pconti31** [comment]
@yesnoj If you edit a theme all newly created UI elements will have your new colors but old ones will not be changed unless you go to Project Options tab and select a change in theme.
Paul--
